### PR TITLE
Unify dataset constructors, simplify dataset building

### PR DESCRIPTION
**Pull request by bpopeters**
This pull request makes a handful of small changes related to the creation of datasets. Most notably, it makes the (hinted at earlier) change of unifying constructors across the various dataset subclasses. It also refactors the use of the various static methods whose purpose was basically to parse corpus files.

### CODE REVIEW

**Comment by vince62s** [comment]
@bpopeters you removed something we introduced with @pltrdy  which is the possibility to build a dataset from an iterator and not from a file.
This is needed especially when we use the server mode,we don't want to create / read files.
We just want to do this on the fly.
If there is a better way to handle this, welcome, but real life is not always files.

**Comment by bpopeters** [comment]
@vince62s does this solve the issues around building iterators from things besides files?

**Comment by vince62s** [comment]
I didn't have time to test today, we'll try tomorrow.

File: onmt/inputters/text_dataset.py [hunk 1]
```diff
@@ -133,14 +71,17 @@ def make_examples(text_iter, truncate, side):
             values are more or less the result of tokenizing with those
             fields.
         """
-        # this function basically reimplements what torchtext Fields and
-        # examples give you for free. It should not exist.
-        for i, line in enumerate(text_iter):
-            line = line.strip().split()
```
Comment (francoishernandez) [inline onmt/inputters/text_dataset.py]: This split() ensured that tokens were handled correctly in the extract_text_features call a few lines down.
Removing it ends up in considering every character as a token, so basically breaking things.
Was there a specific reason for removing it?

**Comment by vince62s** [comment]
also you need to make this change to translation_server.py:
--- a/onmt/translate/translation_server.py
+++ b/onmt/translate/translation_server.py
@@ -372,7 +372,7 @@ class ServerModel:
         if len(texts_to_translate) > 0:
             try:
                 scores, predictions = self.translator.translate(
-                    src_data_iter=texts_to_translate,
+                    texts_to_translate,
                     batch_size=self.opt.batch_size)
             except (RuntimeError, Exception) as e:
                 err = "Error: %s" % str(e)

File: onmt/inputters/text_dataset.py (hunk 1 above)
Comment (bpopeters) [reply onmt/inputters/text_dataset.py]: First, good catch. In this particular case, I moved the `split` to the `_read_file` method and forgot to account for whether the iterator output would be tokenized. I can change it around again so values from an iterator also get split like that.

Second, I am refactoring toward the models being completely flexible with regard to tokenization. Hard-coding `split` into the database code makes it impossible to do that. It also causes the `torchtext.data.Field`'s own tokenizer not to be used, which is counterintuitive if you're familiar with torchtext, nonobvious from looking at the code in OpenNMT (I initially tried specifying the `tokenize` argument when creating the fields and was confused that it appeared not to do anything), and not really documented anywhere. In a different branch with a few commits on top of this, I've been able remove not just the splitting but also `extract_text_features`.

**Comment by bpopeters** [comment]
The build is erroring, not failing.

File: onmt/translate/translator.py [hunk 2]
```diff
@@ -156,25 +151,26 @@ def translate(
...
             sample_rate=self.sample_rate,
             window_size=self.window_size,
             window_stride=self.window_stride,
             window=self.window,
             use_filter_pred=self.use_filter_pred,
-            image_channel_size=self.image_channel_size
+            image_channel_size=self.image_channel_size,
+            dynamic_dict=self.copy_attn
         )
+        print(data.examples[0].__dict__)
```
Comment (francoishernandez) [inline onmt/translate/translator.py]: I think this print is not wanted here.

**Comment by vince62s** [comment]
Thanks!

**Pull request merged**